### PR TITLE
Create cli tool, settings, and slurm utils

### PR DESCRIPTION
**Pull request by m-walters**
I haven't gone through and touched/refactored all the places that these `settings` would apply to (for example dry-run not writing anywhere). So we should patch this in as we go.

### CODE REVIEW

File: settings.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        # Automatically determined by BASE_NAME and RUN_NAME
+        # Should not be a full path
+        self.STORE_PATH = None
+
+        self.DRY_RUN = False
+        self.SEED = None
+        self.VERBOSE = False
+        self.NON_INTERACTIVE = False
+
+        return
```
Comment (mukobi) [inline settings.py:31]: I think we may just not need this class/file. How I tend to prefer to do this kind of thing is with a bunch of parameters defined with argparse, then load them with wandb.init()--see https://docs.wandb.ai/guides/track/config#set-the-configuration-with-argparse. Then, Weights & Biases will handle all the config saving and overriding for us.

**Review by mukobi — CHANGES_REQUESTED** [review]

File: README.md [hunk 2]
```diff
@@ -5,7 +5,42 @@
 1. Install PyTorch separately (for local models). You need torch for most experiments and torchvision for the MNIST toy model.
 > See https://pytorch.org/get-started/locally/
 
-2. Install the requirements
+2. There are two requirements files: `requirements.txt` and `requirements-dev.txt`. The former is pared back and
```
[the remark this answers is not in the document]
Comment (m-walters) [reply README.md:8]: Yeah, the idea is that we don't need to be installing python notebooks and matplotlib etc when creating envs for program execution on cluster.

File: main.py [hunk 3]
```diff
@@ -65,12 +68,34 @@ def calculate_unadaptability_metrics(
...
+            print(f"Creating new Store path dir: {ufm_settings.STORE_PATH}")
+            os.mkdir(ufm_settings.STORE_PATH)
+        else:
+            answer = input(f"Store path '{ufm_settings.STORE_PATH}' does not exist. Create? [y/n]")
+            if answer != 'y':
+                os.mkdir(ufm_settings.STORE_PATH)
+            else:
+                # Exit
+                print("Exiting")
+                sys.exit()
```
[the remark this answers is not in the document]
Comment (m-walters) [reply main.py:98]: Yeah that's what non-interactive was for, to allow a confirmation and clarification around creating/overwriting directories, but i'll scrap it

File: requirements.txt [hunk 4]
```diff
@@ -1,9 +1,5 @@
...
 tqdm
 torch
 torchvision
-ipython
-ipykernel
-ipywidgets
-omegaconf
\ No newline at end of file
+omegaconf
+numpy
```
[the remark this answers is not in the document]
Comment (m-walters) [reply requirements.txt:5]: Will merge when creating `toml` etc.

[the remark this answers is not in the document]
Comment (m-walters) [reply requirements.txt:5]: Alright

File: sbatch_run [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+#SBATCH --time=00:10:00  # time allocation, HH:MM:SS
+#SBATCH --nodes=1
+#SBATCH --partition=single
+#SBATCH --mem=128  # memory per node; default is megabytes; Nvidia A100 has 80G; can use G for gigs etc
```
[the remark this answers is not in the document]
Comment (m-walters) [reply sbatch_run:5]: I left this as small from my testing. `--mem` is memory per "node" which I imagine includes GPUs. Can leave out though, sure

File: settings.py (hunk 1 above)
Comment (m-walters) [reply settings.py:31]: Wouldn't this require passing the config object to basically every function if they want to query any settings/configs? I think Rohan has already set up config parsing with OmegaConf

File: cli.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    )
+    parser.add_argument(
+        "--name", type=str, default=None, help="Optionally identify your run"
+    )
+    # Verbose is boolean for now
+    parser.add_argument(
+        "--verbose", action="store_true", help="Increase output verbosity"
+    )
+    parser.add_argument(
+        "--non-interactive", action="store_true", help="Whether to run non-interactively"
```
[the remark this answers is not in the document]
Comment (m-walters) [reply cli.py:33]: Removed non-interactive

File: README.md [hunk 7]
```diff
@@ -5,7 +5,42 @@
...
+
+### Setup script
+
+The script `setup.sh` sets up the environment for running jobs. It should be called in the sbatch scripts. Of note it:
+
+- Enters or initializes a `venv` and pip installs packages.
+  - **Note:** You can speed up this process by doing these steps in your working directory and the slurm jobs will 
+    use them instead of redoing it all. (At least _I think_ it will. It might get weird if you're using multiple 
+    nodes or something. Regardless, at worst it just initializes these things from scratch).
+- Updates the `PYTHONPATH` to include the working directory
```
[the remark this answers is not in the document]
Comment (m-walters) [reply README.md:21]: hmm, I'd rather not hold up the PR configuring this as a package, so I'll remove this and we'll avoid nested imports until then

File: settings.py (hunk 1 above)
Comment (mukobi) [reply settings.py]: `wandb.config` is a dict available globally, so we don't even have to pass it in. See https://docs.wandb.ai/guides/track/config

File: README.md [hunk 8]
```diff
@@ -5,7 +5,42 @@
...
+have a simple executable `sbatch_run` you can reference.
+
+You can then view your jobs with `squeue`. When plain, this shows all jobs, but you can isolate your own with 
+`squeue -u <username>`.
+
+### Setup script
+
+The script `setup.sh` sets up the environment for running jobs. It should be called in the sbatch scripts. Of note it:
+
+- Enters or initializes a `venv` and pip installs packages.
```
[the remark this answers is not in the document]
Comment (mukobi) [reply README.md:17]: Or since these are just for initial setup, you could have two separate setup scripts for conda and venv. Though that feels needlessly redundant.

File: sbatch_run (hunk 5 above)
[the remark this answers is not in the document]
Comment (mukobi) [reply sbatch_run]: Node kinda means an individual computer. A node can multiple GPUs and CPUs attached to it. --mem still specifies the CPU RAM, in MB if you omit a unit. https://slurm.schedmd.com/srun.html#:~:text=to%20job%20allocations.-,%2D%2Dmem%3D%3Csize%3E%5Bunits%5D,-Specify%20the%20real Should just remove for now.

[the remark this answers is not in the document]
Comment (m-walters) [reply sbatch_run]: Sure, np!

File: README.md [hunk 9]
```diff
@@ -2,10 +2,14 @@
...
+You can install in dev mode to include additional dependencies for data visualization etc.
 
-2. Install the requirements
 ```bash
-pip install -r requirements.txt
-```
\ No newline at end of file
+pip install ".[dev]"
+# Or for editable mode
+pip install -e ".[dev]"
```
Comment (mukobi) [inline README.md:14]: You probably should tell developers to install only in editable mode, so you can remove the first one here.

File: cli.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    parser.add_argument(
+        "--name", type=str, default=None, help="Optionally identify your run"
+    )
+    # Verbose is boolean for now
+    parser.add_argument(
+        "--verbose", action="store_true", help="Increase output verbosity"
+    )
+
+    args = parser.parse_args()
+    config = load_config(args.config)
```
Comment (mukobi) [inline cli.py:34]: Since we'll be trying many things, I'm not sure it makes sense to have many settings stored in config files. Would recommend not using OmegaConfig, instead defining all our arguments here in argparse. Can make them hierarchical, like `--model.path` and `--model.lora.rank`. Then, when loaded into `wandb.init(config=args)`, `wandb.config` will be a hierarchical object of all the settings. That also makes wandb sweeps over these settings nicer to work with.

File: settings.py (hunk 1 above)
Comment (m-walters) [reply settings.py]: ok added

File: cli.py (hunk 10 above)
Comment (m-walters) [reply cli.py:46]: But we define all out model parameters in these config files. Putting these all in through argv sounds super painful

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (m-walters) [reply README.md]: How about we leave this for a later PR and get this one merging in.

File: cli.py (hunk 10 above)
Comment (mukobi) [reply cli.py:46]: If using config files, how do you specify changes for individual experiments (surely, we don't want a config file per experimental condition)? We'd set default values in cli.py for each of the values, so you'd only specify the things you want to be different from default. This makes it especially nice to run sweeps over config values https://docs.wandb.ai/guides/sweeps/walkthrough

Comment (m-walters) [reply cli.py:46]: Hmm..I'm not aware of a way to, through argparse, define an argument such as `"--model"` that adapts to users passing in new tokens like `--model.path=...`. Instead of spelling out every potential param, I have a `json` solution where `"--model"` is a string arg, but users can make the string a json object e.g. `"{'path': '/some/path'}"` and this will be loaded as a dict to conf.

It's worth noting that you can specify wandb sweeps as config files: https://docs.wandb.ai/guides/sweeps/define-sweep-configuration
so I think we should keep config loading as well.

Comment (mukobi) [reply cli.py:55]: Argparse doesn't adapt to new sequences of parameters, but we don't need that functionality. We'd just specify each of the configurable parameters to our experiments exactly once with default values and tooltips in cli.py. Using periods in the parameter values would be purely aesthetic to help separate different clusters of parameters from each other, though we might use some other syntax like underscores.

I'm down to try omegaconf, but it feels messy to have both that kind of a config file and other argparsed args. If you want to use omegaconf files, it might be better to use [hydra](https://hydra.cc/docs/intro/) to automatically merge in the default values from a config file and any updated values from the command line or a sweep. Hydra is kinda officially supported by wandb if you explicitly convert the config to a dict before wandb project initialization (https://docs.wandb.ai/guides/integrations/hydra). Otherwise, you can use OmegaConf.load(), OmegaConf.from_cli(), and OmegaConf.merge() together to implement hydra-like behavior of having default config file values and updated cli values.

File: configs/wandb_sample.yaml
Comment (mukobi) [inline configs/wandb_sample.yaml:1]: We probably shouldn't commit this example file to our repo. Unless this was accidentally included?

Also, I think it could be good to have our config file(s) and sweep files separate, e.g. in folders called `configs/` and `sweeps/`

File: cli.py [hunk 11]
```diff
@@ -75,5 +84,13 @@ def parse_arguments() -> Tuple[Any, DictConfig]:
...
         "seed": args.seed if args.seed else config.get("seed", 8675309),
     })
+    
+    # Overrides for json objects
+    if args.model:
+        config.update({"model": json.loads(args.model)})
+    if args.finetune:
+        config.update({"finetune": json.loads(args.finetune)})
+    if args.pretrain:
+        config.update({"pretrain": json.loads(args.pretrain)})
```
Comment (mukobi) [inline cli.py:94]: This config overriding will get annoying. Would suggest Hydra, omegaconf merging, or argparse default values. See [this other comment](https://github.com/mukobi/Unadaptable-Foundation-Models/pull/16/commits/36d103577ff5072b0f25294c545d2dff2e0b56de#r1558812665)

**Review by mukobi — CHANGES_REQUESTED** [review]
Just the omegaconf config parsing needs sorting out, then lgtm

File: README.md [hunk 12]
```diff
@@ -13,10 +13,147 @@ pip install -e ".[dev]"
...
+```
+
+wraps around our `main` function and seamlessly allows for config parameter changes
+"on-the-fly" as arguments to `main.py`, for example
+
+```bash
+python main.py pretrain.epochs=20
+```
+
+overwrites the `pretrain.epochs` parameter. Notice that `pretrain.epochs` was specified to ArgParse in advance.
```
Comment (mukobi) [inline README.md:48]: prob shouldn't say ArgParse here

File: README.md [hunk 13]
```diff
@@ -13,10 +13,147 @@ pip install -e ".[dev]"
...
+the command line
+```bash
+python main.py --config-name=mnist pretrain.epochs=20 seed=8675309
+# Selecting a different unadapt method
+python main.py --config-name=mnist unadapt=prune
+```
+You can also provide one-off params with `+`:
+```bash
+python main.py --config-name=mnist +slurm_log=/slurm/path
+```
```
Comment (mukobi) [inline README.md:138]: Are we ever gonna use one-off params? Possibly better to just require everything be defined in the config already, then we don't have to check if things are in the config first. Maybe the exception to this could be boolean flags we want to easily enable, like `+disablewandb`, but you could also just have that explicit in the configs as `wandb: enabled` and override it with `wandb=disabled`

File: config.py [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+def parse_config(cfg: DictConfig) -> DictConfig:
+    """
+    Parse the config compiled by hydra
+    """
+    print("Initializing run. Config parameters:")
+    # Tags must be list if provided
+    if isinstance(cfg.get("tags", None), str):
+        cfg["tags"] = [cfg["tags"]]
+
+    print(OmegaConf.to_yaml(cfg))
```
Comment (mukobi) [inline config.py:13]: Eventually, we should probably use a logger passed in from main so we can filter and such. Should be fine for now.

File: main.py [hunk 15]
```diff
@@ -66,42 +69,56 @@ def calculate_unadaptability_metrics(
...
+    seed_all(cfg.seed)
     device = "cuda" if torch.cuda.is_available() else "cpu"
     print(f"Using {device} device")
 
-    model = get_model(config.model, device)
+    return
+
+# For recording tracebacks on CAIS:
+# https://cluster.safe.ai/#enabling-debugging-for-distributed-training
+@record
```
Comment (mukobi) [inline main.py:94]: We prob won't need distributed training. Would suggest leaving this out until we prove that we need it.

File: README.md (hunk 13 above)
Comment (m-walters) [reply README.md:138]: I was just describing built-in hydra functionality. We can't stop users from doing it, but i'll remove
